rackcity/views/decommission_views.py

```diff
@@ -41,33 +48,104 @@ def decommission_asset(request):
             status=HTTPStatus.BAD_REQUEST
         )
     id = data['id']
-    try:
-        asset = Asset.objects.get(id=id)
-    except Asset.DoesNotExist:
-        return JsonResponse(
-            {
-                "failure_message":
-                    Status.ERROR.value +
-                    "Asset" + GenericFailure.DOES_NOT_EXIST.value,
-                "errors": "No existing asset with id="+str(id)
-            },
-            status=HTTPStatus.BAD_REQUEST
-        )
-    if not user_has_asset_permission(
-        request.user,
-        asset.rack.datacenter
-    ):
+    decommissioned_asset_cp = None
+    if change_plan:
+        if not AssetCP.objects.filter(id=id).exists() and Asset.objects.filter(id=id).exists():
+            existing_asset = Asset.objects.get(id=id)
+            decommissioned_asset_cp = AssetCP(
+                change_plan=change_plan,
+                related_asset=existing_asset,
+                is_decommissioned=True) 
+            for field in existing_asset._meta.fields:
+                if not (field.name == "id" or field.name == "assetid_ptr"):
+                    setattr(decommissioned_asset_cp, field.name, getattr(
+                        existing_asset, field.name))
+        else:
+            try:
+                decommissioned_asset_cp = AssetCP.objects.get(id=id)
+                decommissioned_asset_cp.is_decommissioned=True
+
+            except Asset.DoesNotExist:
+                return JsonResponse(
+                    {
+                        "failure_message":
+                            Status.ERROR.value +
+                            "Asset" + GenericFailure.DOES_NOT_EXIST.value,
+                        "errors": "No existing asset in change plan with id="+str(id)
+                    },
+                    status=HTTPStatus.BAD_REQUEST
+                )
+        if not user_has_asset_permission(
+            request.user,
+            decomissioned_asset_cp.rack.datacenter
```

Comment:
typo - `decomissioned`

Reply:
omg thank god i did not work on decommissioning